src/core/load-fragment.c

```diff
@@ -5421,6 +5422,56 @@ int config_parse_ip_filter_bpf_progs(
         return 0;
 }
 
+int config_parse_bpf_foreign_program(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        _cleanup_free_ char *bpffs_path = NULL;
+        _cleanup_free_ char *resolved = NULL;
+        enum bpf_attach_type attach_type;
+        Unit *u = userdata;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+
+        if (isempty(rvalue)) {
+                bpf_foreign_reset(u);
+                return 0;
+        }
+
+        r = unit_full_printf(u, rvalue, &resolved);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to resolve unit specifiers in '%s', ignoring: %m", rvalue);
+                return 0;
+        }
+
+        r = bpf_foreign_program_from_string(resolved, &attach_type, &bpffs_path);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                                "Failed to resolve foreign BPF program for path='%s', ignoring: %m", resolved);
+                return 0;
+        }
+
+        r = path_simplify_and_warn(bpffs_path, PATH_CHECK_ABSOLUTE, unit, filename, line, lvalue);
+        if (r < 0)
+                return 0;
+
+        r = bpf_foreign_prepare(u, attach_type, bpffs_path);
```

Comment:
so, this means you ref the bpf program at the moment systemd configuration is loaded, which can be super-duper early during boot, i.e. where it is highly unlikely the program was already loaded into the kernel yet.

Generally: when we load config we should only log config, i.e. not actually statr referencing runtime objects. Hence, I think the bpf programs should be reference right before we start using them, and until then we should only track to the path to them. That would mean that the programs that upload the bpf programs into the kernel can run as systemd service, and just place themselves with Before= before the service that actually shall use them

Reply:
@poettering 

I misunderstood some comments from the previous review, removed paths completely and moved loading to  that early stage.
Move loading and referencing back to `cgroup_context_apply`.